modules/mod_ginger_activity/mod_ginger_activity.erl

```diff
@@ -85,28 +99,46 @@ register_activity(RscId, Context) ->
                     z_convert:to_binary(string:strip(lists:last(string:tokens(Hosts, ","))))
             end
     end,
-    Entry = #entry{rsc_id = RscId, time = Time, user_id = UserId, ip_address = IpAddress},
-    z_notifier:notify({ginger_activity, Entry}, Context),
+    Activity = #ginger_activity{
+        rsc_id = RscId,
+        target_id = Target,
+        time = Time,
+        user_id = UserId,
+        ip_address = IpAddress
+    },
+    z_notifier:notify({ginger_activity, Activity}, Context),
     ok.
 
-pid_observe_ginger_activity(_Pid, {ginger_activity, Entry}, Context) ->
-    #entry{rsc_id = RscId, time = Time, user_id = UserId, ip_address = IpAddress} = Entry,
+pid_observe_ginger_activity(_Pid, {ginger_activity, Activity}, Context) ->
+    #ginger_activity{rsc_id = RscId} = Activity,
     case z_convert:to_bool(m_config:get_value(mod_ginger_activity, persist_activity, Context)) of
         true ->
-            insert_activity(RscId, Time, UserId, IpAddress, Context),
+            insert_activity(Activity, Context),
             z_pivot_rsc:pivot(RscId, Context),
             ok;
         false ->
             ok
     end.
 
-% @doc single entry point for inserting activity into the database
-% insert_activity(RscId, Context) ->
-%     insert_activity(RscId, calendar:local_time(), Context).
-% insert_activity(RscId, DateTime, Context) ->
-%     insert_activity(RscId, DateTime, undefined, Context).
-% insert_activity(RscId, DateTime, UserId, Context) ->
-%     insert_activity(RscId, DateTime, UserId, undefined, Context).
-insert_activity(RscId, DateTime, UserId, IpAddress, Context) ->
-    Props = [{rsc_id, RscId}, {time, DateTime}, {user_id, UserId}, {ip_address, IpAddress}],
-    z_db:insert(activity_log, Props, Context).
+-spec insert_activity(activity(), z:context()) -> ok.
+insert_activity(Activity, Context) ->
+    #ginger_activity{
+        rsc_id = Rsc,
+        target_id = Target,
```

Comment:
target_id might be a confusing name, though I think I don't know a better alternative.
The target is an optional id of e.g. the page where resource Rsc was created. It provides a link to the context of Rsc which might be more useful to navigate to (instead of to the page for Rsc itself) when clicking the activity notification. So the "best" alternative I can think of would be "context_rsc_id". In its defense, the concept of "context of the activity" seems quite natural and useful.